From: =?utf-8?q?Florian_M=C3=BCllner?= <fmuellner@gnome.org>
Date: Wed, 1 Apr 2020 21:04:03 +0200
Subject: extensionSystem: Add method for opening extension prefs

Extension that want to expose their own preferences (for example as menu
items) do that by passing their UUID to gnome-shell-extension-prefs.

But since 3.36.1 the app is optional and no longer accepts arguments on
the command line. To adjust, extensions now need to make a D-Bus call
the extensions portal, just like the app and gnome-shell.

We will add a convenience method for that purpose, so it makes
sense to share the existing code. As it's extension-related, the
extension manager looks like the right place ...

https://gitlab.gnome.org/GNOME/gnome-shell/-/merge_requests/1163

(cherry picked from commit 45bc850715f875aca705ddccc39b9cc46e757bd1)

Origin: upstream, 3.36.2, commit:efee3aa7495a33ca68ed02ed3089910f66622d96
---
 js/ui/extensionSystem.js | 19 +++++++++++++++++++
 js/ui/shellDBus.js       | 12 +-----------
 2 files changed, 20 insertions(+), 11 deletions(-)

diff --git a/js/ui/extensionSystem.js b/js/ui/extensionSystem.js
index 4fab0c3..21ee8a1 100644
--- a/js/ui/extensionSystem.js
+++ b/js/ui/extensionSystem.js
@@ -215,6 +215,25 @@ var ExtensionManager = class {
         return true;
     }
 
+    openExtensionPrefs(uuid, parentWindow, options) {
+        const extension = this.lookup(uuid);
+        if (!extension || !extension.hasPrefs)
+            return false;
+
+        Gio.DBus.session.call(
+            'org.gnome.Shell.Extensions',
+            '/org/gnome/Shell/Extensions',
+            'org.gnome.Shell.Extensions',
+            'OpenExtensionPrefs',
+            new GLib.Variant('(ssa{sv})', [uuid, parentWindow, options]),
+            null,
+            Gio.DBusCallFlags.NONE,
+            -1,
+            null,
+            (conn, res) => conn.call_finish(res));
+        return true;
+    }
+
     notifyExtensionUpdate(uuid) {
         let extension = this.lookup(uuid);
         if (!extension)
diff --git a/js/ui/shellDBus.js b/js/ui/shellDBus.js
index f64081b..75a2f34 100644
--- a/js/ui/shellDBus.js
+++ b/js/ui/shellDBus.js
@@ -316,17 +316,7 @@ var GnomeShellExtensions = class {
     }
 
     OpenExtensionPrefs(uuid, parentWindow, options) {
-        Gio.DBus.session.call(
-            'org.gnome.Shell.Extensions',
-            '/org/gnome/Shell/Extensions',
-            'org.gnome.Shell.Extensions',
-            'OpenExtensionPrefs',
-            new GLib.Variant('(ssa{sv})', [uuid, parentWindow, options]),
-            null,
-            Gio.DBusCallFlags.NONE,
-            -1,
-            null,
-            (conn, res) => conn.call_finish(res));
+        Main.extensionManager.openExtensionPrefs(uuid, parentWindow, options);
     }
 
     ReloadExtensionAsync(params, invocation) {
